cupyx/scipy/linalg/special_matrices.py

```diff
@@ -0,0 +1,552 @@
+                      strides=(-n, n)).copy()
+
+
+def circulant(c):
+    """Construct a circulant matrix.
+
+    Args:
+        c (cupy.ndarray): 1-D array, the first column of the matrix.
+
+    Returns:
+        cupy.ndarray: A circulant matrix whose first column is ``c``.
+
+    .. seealso:: :func:`cupyx.scipy.linalg.toeplitz`
+    .. seealso:: :func:`cupyx.scipy.linalg.hankel`
+    .. seealso:: :func:`cupyx.scipy.linalg.solve_circulant`
+    .. seealso:: :func:`cupyx.scipy.linalg.fiedler`
+    .. seealso:: :func:`scipy.linalg.circulant`
+    """
+    c = c.ravel()
+    # Form an extended array that could be strided to give circulant version
+    c_ext = cupy.concatenate((c[::-1], c[:0:-1]))
+    L = c.size
+    n = c_ext.strides[0]
+    return as_strided(c_ext[L-1:], shape=(L, L), strides=(-n, n)).copy()
+
+
+def hankel(c, r=None):
+    """Construct a Hankel matrix.
+
+    The Hankel matrix has constant anti-diagonals, with ``c`` as its first
+    column and ``r`` as its last row. If ``r`` is not given, then
+    ``r = zeros_like(c)`` is assumed.
+
+    Args:
+        c (cupy.ndarray): First column of the matrix. Whatever the actual shape
+            of ``c``, it will be converted to a 1-D array.
+        r (cupy.ndarray, optionnal): Last row of the matrix. If None,
+            ``r = zeros_like(c)`` is assumed. ``r[0]`` is ignored; the last row
+            of the returned matrix is ``[c[-1], r[1:]]``. Whatever the actual
+            shape of ``r``, it will be converted to a 1-D array.
+
+    Returns:
+        cupy.ndarray: The Hankel matrix. Dtype is the same as
+            ``(c[0] + r[0]).dtype``.
+
+    .. seealso:: :func:`cupyx.scipy.linalg.toeplitz`
+    .. seealso:: :func:`cupyx.scipy.linalg.circulant`
+    .. seealso:: :func:`scipy.linalg.hankel`
+    """
+    c = c.ravel()
+    r = cupy.zeros_like(c) if r is None else r.ravel()
+    # Form a 1-D array of values to be used in the matrix, containing `c`
+    # followed by r[1:].
+    vals = cupy.concatenate((c, r[1:]))
+    # Stride on concatenated array to get hankel matrix
+    n = vals.strides[0]
+    return as_strided(vals, shape=(c.size, r.size), strides=(n, n)).copy()
+
+
+def hadamard(n, dtype=int):
+    """Construct an Hadamard matrix.
+
+    Constructs an n-by-n Hadamard matrix, using Sylvester's construction. ``n``
+    must be a power of 2.
+
+    Args:
+        n (int): The order of the matrix. ``n`` must be a power of 2.
+        dtype (dtype, optional): The data type of the array to be constructed.
+
+    Returns:
+        (cupy.ndarray): The Hadamard matrix.
+
+    .. seealso:: :func:`scipy.linalg.hadamard`
+    """
+    lg2 = 0 if n < 1 else (int(n).bit_length() - 1)
+    if 2 ** lg2 != n:
+        raise ValueError("n must be an positive a power of 2 integer")
+
+    # Sylvester's construction
+    H = cupy.ones((1, 1), dtype=dtype)
+    for _ in range(0, lg2):
+        n = H.shape[0]
+        H = cupy.tile(H, (2, 2))
+        H[n:, n:] *= -1
```

Comment:
Tried figuring it out but couldn't do it easily. Likely for each value in the matrix you would have to figure out the number of times it would be negated. The lower-right corner is negated each iteration (after tiling).

Reply:
I actually found a clever way to do this one very quickly with a raw kernel. Each value is positive or negative one. The negative values are when the number of bits in `i & j` is odd where `i` is the row and `j` is the column (starting both at 0). After a bit more testing I will commit that version as well.